### PR TITLE
Add a cloud event helper

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Add a cloud event helper that is going to be used by the
CloudEventPipelineResource to send cloud events.
This includes an initial list of the type of events that can be
generated by Tekton.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) and [TaskRun](../tekton/publish-run.yaml) to build and release this image

### CODE REVIEW

File: pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent_test.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+limitations under the License.
+*/
+
+package cloudevent
+
+import (
+  "encoding/json"
+  "fmt"
+  "regexp"
+	"testing"
```
Comment (dibyom) [inline pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent_test.go]: Indenting seems off here.

File: pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudeventclient.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudevent
+
+import (
+	"context"
+
+  "github.com/knative/eventing-sources/pkg/kncloudevents"
```
Comment (dibyom) [inline pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudeventclient.go]: Missing some indentation here too.

Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudeventclient.go]: Thanks! I ran the whole module through go fmt now.

File: pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent_test.go (hunk 1 above)
Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/cloudevent/cloudevent_test.go]: Thanks! I ran the whole module through go fmt now.

**Comment by afrittoli** [comment]
@akihikokuroda I want to address your comments on https://github.com/tektoncd/pipeline/pull/837, but I didn't get a chance yet since I was traveling. 
The changes to be made are:
- move TaskRun from the root of the event body to a TaskRun field, so that it's easier to add data in future if needed
- add versioning to the cloudevent headers

Since this one has already been reviewed, I could address your comments on a PR on top. Until https://github.com/tektoncd/pipeline/pull/837 merges the resource won't be actually usable. 
@dlorenc thoughs?

**Comment by bobcatfish** [comment]
Weeeird seems like a pretty similar error both times:

```
I0723 03:06:20.887] --- FAIL: TestTaskRunPipelineRunCancel (2.12s)
I0723 03:06:20.888]     --- FAIL: TestTaskRunPipelineRunCancel/With_retries (3.82s)
I0723 03:06:20.888]         init_test.go:111: Create namespace arendelle-5kh2p to deploy to
I0723 03:06:20.888]         init_test.go:123: Verify SA "default" is created in namespace "arendelle-5kh2p"
I0723 03:06:20.889]         cancel_test.go:66: Creating Task in namespace arendelle-5kh2p
I0723 03:06:20.889]         cancel_test.go:74: Creating Pipeline in namespace arendelle-5kh2p
I0723 03:06:20.889]         cancel_test.go:84: Creating PipelineRun in namespace arendelle-5kh2p
I0723 03:06:20.889]         cancel_test.go:89: Waiting for Pipelinerun pear in namespace arendelle-5kh2p to be started
I0723 03:06:20.889]         cancel_test.go:110: Waiting for TaskRuns from PipelineRun pear in namespace arendelle-5kh2p to be running
I0723 03:06:20.890]         cancel_test.go:139: Failed to cancel PipelineRun `pear`: Operation cannot be fulfilled on pipelineruns.tekton.dev "pear": the object has been modified; please apply your changes to the latest version and try again
```

But the logic this PR is adding isn't being invoked yet so it seems _pretty_ unlikely these changes are the cause! 🤔

**Comment by afrittoli** [comment]
> Weeeird seems like a pretty similar error both times:
> 
> ```
> I0723 03:06:20.887] --- FAIL: TestTaskRunPipelineRunCancel (2.12s)
> I0723 03:06:20.888]     --- FAIL: TestTaskRunPipelineRunCancel/With_retries (3.82s)
> I0723 03:06:20.888]         init_test.go:111: Create namespace arendelle-5kh2p to deploy to
> I0723 03:06:20.888]         init_test.go:123: Verify SA "default" is created in namespace "arendelle-5kh2p"
> I0723 03:06:20.889]         cancel_test.go:66: Creating Task in namespace arendelle-5kh2p
> I0723 03:06:20.889]         cancel_test.go:74: Creating Pipeline in namespace arendelle-5kh2p
> I0723 03:06:20.889]         cancel_test.go:84: Creating PipelineRun in namespace arendelle-5kh2p
> I0723 03:06:20.889]         cancel_test.go:89: Waiting for Pipelinerun pear in namespace arendelle-5kh2p to be started
> I0723 03:06:20.889]         cancel_test.go:110: Waiting for TaskRuns from PipelineRun pear in namespace arendelle-5kh2p to be running
> I0723 03:06:20.890]         cancel_test.go:139: Failed to cancel PipelineRun `pear`: Operation cannot be fulfilled on pipelineruns.tekton.dev "pear": the object has been modified; please apply your changes to the latest version and try again
> ```
> 
> But the logic this PR is adding isn't being invoked yet so it seems _pretty_ unlikely these changes are the cause! 🤔

That was my though exactly. There seems to be some kind of race condition happening. Having test results data with test granularity (I think we only do it on test job granularity today) might be helpful to troubleshoot this; I look into it.

**Pull request merged**